src/restruct_poc/solver/constraint.hpp

```diff
@@ -50,29 +55,33 @@ struct Constraint {
             this->X0[2] = this->target_node.x[2] - this->base_node.x[2];
 
             // If rotation control constraint, vec is rotation axis
-            if (constraint_type == ConstraintType::kRotationControl) {
-                this->x_axis = vec;
-            } else if (constraint_type == ConstraintType::kCylindrical) {
-                Array_3 x_hat = this->X0;
-                auto length = sqrt(x_hat[0] * x_hat[0] + x_hat[1] * x_hat[1] + x_hat[2] * x_hat[2]);
-                if (length > 1.0e-10) {
-                    x_hat[0] /= length;
-                    x_hat[1] /= length;
-                    x_hat[2] /= length;
-                } else {
-                    x_hat = {1., 0., 0.};
-                }
-                Array_3 x = {1., 0., 0.};
+            if (constraint_type == ConstraintType::kCylindrical) {
+                Array_3 x{1., 0., 0.};
+                auto x_hat = UnitVector(this->X0);
 
+                // Create rotation matrix which rotates x to match vector
                 auto v = CrossProduct(x, x_hat);
-                auto c = x[0] * x_hat[0] + x[1] * x_hat[1] + x[2] * x_hat[2];
+                auto c = DotProduct(x_hat, x);
                 auto k = 1. / (1. + c);
+                Array_3x3 R = {
+                    {{
+                         v[0] * v[0] * k + c,
+                         v[0] * v[1] * k - v[2],
+                         v[0] * v[2] * k + v[1],
+                     },
+                     {
+                         v[1] * v[0] * k + v[2],
+                         v[1] * v[1] * k + c,
+                         v[1] * v[2] * k - v[0],
+                     },
+                     {
+                         v[2] * v[0] * k - v[1],
+                         v[2] * v[1] * k + v[0],
+                         v[2] * v[2] * k + c,
+                     }}};
 
-                std::array<std::array<double, 3>, 3> R = {
-                    {{v[0] * v[0] * k + c, v[0] * v[1] * k - v[2], v[0] * v[2] * k + v[1]},
-                     {v[1] * v[0] * k + v[2], v[1] * v[1] * k + c, v[1] * v[2] * k - v[0]},
-                     {v[2] * v[0] * k - v[1], v[2] * v[1] * k + v[0], v[2] * v[2] * k + c}}};
-
+                // Columns of rotation matrix are orthogonal unit vectors with
+                // the first column matching the target vector
                 this->x_axis = {R[0][0], R[1][0], R[2][0]};
                 this->y_axis = {R[0][1], R[1][1], R[2][1]};
                 this->z_axis = {R[0][2], R[1][2], R[2][2]};
```

Comment:
Would the following make things a little more readable?

```
// If fixed BC or prescribed displacement, X0 is based on reference position vector
        if (constraint_type == ConstraintType::kFixedBC ||
            constraint_type == ConstraintType::kPrescribedBC) {
            this->X0[0] = this->target_node.x[0] - vec[0];
            this->X0[1] = this->target_node.x[1] - vec[1];
            this->X0[2] = this->target_node.x[2] - vec[2];
            return;
        }

        // Calculate initial difference in position between nodes
        this->X0[0] = this->target_node.x[0] - this->base_node.x[0];
        this->X0[1] = this->target_node.x[1] - this->base_node.x[1];
        this->X0[2] = this->target_node.x[2] - this->base_node.x[2];

        // If rotation control constraint, vec is rotation axis
        if (constraint_type == ConstraintType::kCylindrical) {
            Array_3 x{1., 0., 0.};
            auto x_hat = UnitVector(this->X0);

            // Create rotation matrix which rotates x to match vector
            auto v = CrossProduct(x, x_hat);
            auto c = DotProduct(x_hat, x);
            auto k = 1. / (1. + c);
            Array_3x3 R = {
                {{
                     v[0] * v[0] * k + c,
                     v[0] * v[1] * k - v[2],
                     v[0] * v[2] * k + v[1],
                 },
                 {
                     v[1] * v[0] * k + v[2],
                     v[1] * v[1] * k + c,
                     v[1] * v[2] * k - v[0],
                 },
                 {
                     v[2] * v[0] * k - v[1],
                     v[2] * v[1] * k + v[0],
                     v[2] * v[2] * k + c,
                 }}};

            // Columns of rotation matrix are orthogonal unit vectors with
            // the first column matching the target vector
            this->x_axis = {R[0][0], R[1][0], R[2][0]};
            this->y_axis = {R[0][1], R[1][1], R[2][1]};
            this->z_axis = {R[0][2], R[1][2], R[2][2]};
        }
    }
```

Reply:
Yes, that is cleaner, I'll change it